### PR TITLE
node-labeller: enable node-labeller for ARM64 clusters

### PR DESCRIPTION
**Pull request by dasionov**
This change enables the `node-labeller` routine for ARM64-based clusters in KubeVirt by updating the `archLabellerARM64` structure to drop `shouldLabelNodes()` since it returns true for all arches.
Previously, the `node-labeller` was disabled for ARM64 clusters due to incomplete support for labels like CPU features and models, which are supported on other architectures such as `amd64` and `s390x` - ref https://issues.redhat.com/browse/RHEL-88219.

For this PR, we’ve chosen to enable only machine-type labels for ARM64, as these are critical for virtual machine (VM) scheduling and represent the minimum functionality required from the `node-labeller`.

Note that machine-type labels were not required on arm64 clusters previously because VM scheduling on ARM64 did not depend on them. This changed after #13690 was merged, introducing functionality that relies on machine-type labels for proper VM scheduling.

Machine types are retrieved from the libvirt capabilities.xml, which is reliably available on ARM64 nodes. These values can be safely processed by the node-labeller without introducing risk or other issues.

Other labels supported by `node-labeller` on `amd64` and `s390x` (listed down) are intentionally ignored for ARM64 to minimize risk and focus on essential functionality. Support for additional labels may be added in future PRs.

Currently, the only supported label on ARM64 is:

- SupportedMachineTypeLabel

The following labels are not supported for ARM64 in this PR:
- CPUFeatureLabel
- CPUModelLabel
- CPUModelVendorLabel
- SupportedHostModelMigrationCPU
- CPUTimerLabel
- HypervLabel
- RealtimeLabel
- SEVLabel
- SEVESLabel
- HostModelCPULabel
- HostModelRequiredFeaturesLabel
- NodeHostModelIsObsoleteLabel

Fixes #

### Release note
```release-note
Enable node-labeller for ARM64 clusters, supporting machine-type labels.
```

### CODE REVIEW

**Review by fossedihelm — COMMENTED** [review]
I understood that the node labeller was not relevant to arm.
Can you point out the reasoning behind this?
Thank you

**Comment by dollierp** [comment]
> I understood that the node labeller was not relevant to arm. Can you point out the reasoning behind this? Thank you

The label `machine-type.node.kubevirt.io/virt: "true"` is required since https://github.com/kubevirt/kubevirt/pull/13690.

**Comment by dasionov** [comment]
> I understood that the node labeller was not relevant to arm. Can you point out the reasoning behind this? Thank you

As @dollierp mentioned, #13690 introduces a new label that must exist on nodes to allow scheduling. However, the entire idea was based on a situation that arose even before its implementation, which was aimed at addressing a problem with CPU host-model migrations introduced in #7672.

I find it strange that the changes introduced in #7672 didn’t require the node labeller to be active for ARM clusters.
Is it because the changes only affect the node selector of a target pod during migration?
And since migrations aren’t supported on ARM clusters, we simply didn’t notice until now?

File: pkg/virt-handler/node-labeller/node_labeller_test.go [hunk 1]
```diff
@@ -294,7 +294,7 @@ var _ = Describe("Node-labeller ", func() {
 		Expect(nlController.ShouldLabelNodes()).To(Equal(shouldLabel))
 	},
 		Entry(amd64, amd64, true),
-		Entry(arm64, arm64, false),
+		Entry(arm64, arm64, true),
```
Comment (dankenigsberg) [inline pkg/virt-handler/node-labeller/node_labeller_test.go:297]: now that it returns `true` for all archs, do we need to keep `ShouldLabelNodes()` ?

Comment (dasionov) [reply pkg/virt-handler/node-labeller/node_labeller_test.go:297]: Good point! 
IMO, we can remove the `shouldLabelNodes()` method.
However, I'd appreciate hearing other perspectives on this.
If there are no strong objections, I'd be happy to proceed with the change in a follow-up PR.

Comment (dankenigsberg) [reply pkg/virt-handler/node-labeller/node_labeller_test.go:297]: The only objection is if NodeLabeller cannot run on ARM (or on any other architecture).
We heard rumors that it was failing to run there several years ago due to a libvirt deficiency. But if this is no longer true, let us iron out this wrinkle.

Comment (fossedihelm) [reply pkg/virt-handler/node-labeller/node_labeller_test.go:297]: > I don't see why. If it's not used, let's remove it.

+1, in case we need it in the future, it will be reintroduced.

> do we know if node-labeller is running well on ARM?

@zhlhahaha Can you help here? Thanks

Comment (dasionov) [reply pkg/virt-handler/node-labeller/node_labeller_test.go:297]: Sorry for not specifying this explicitly earlier.
The changes I made were tested manually on an ARM64 cluster (not a heterogeneous one).

The main issue was that hostCpuModel.Name is empty on ARM.
Most of the existing code handled this well — for example:
```
for key := range cpuFeatures {
    newLabels[kubevirtv1.CPUFeatureLabel+key] = "true"
}
```
On ARM, cpuFeatures is empty, so the for loop is naturally skipped without any problems.

However, for hostCpuModel.Name, the issue was that an empty string was misinterpreted as a valid CPU host model.
For example, in this snippet:
```
if _, hostModelObsolete := obsoleteCPUsx86[hostCpuModel.Name]; !hostModelObsolete {
	newLabels[kubevirtv1.SupportedHostModelMigrationCPU+hostCpuModel.Name] = "true"
}
```
We check if the host CPU model is obsolete by looking it up in a known list.
Since hostCpuModel.Name is empty on ARM, it is not found in the list, and incorrectly treated as a valid CPU model.
This leads to a bug where an invalid label is created.
Initially, I tried fixing this by wrapping each case individually with:
```
if hostCpuModel.Name != "" {
    ...
}
```
However, since there were four such cases, I decided to instead group all of them under a single:
```
if n.arch.supportsHostModel() {
    ...
}
```

**Comment by andreabolognani** [comment]
@dasionov my reading of your changes is that you've made a bunch of code dealing with CPU models conditional on `arch.supportsHostModel()`, so it will not execute on aarch64. That seems sensible. Just for completeness' sake, though people are probably already aware of it, only `<cpu mode='host-passthrough'/>` is currently usable on aarch64. Work is ongoing to lift that limitation, but it will likely take a fairly long time before that will happen.

**Review by xpivarc — COMMENTED** [review]
Let's split the semantic changes to commits. The cpu model should be unit testable. Please add a test that demonstrate the problem of empty host model.

**Comment by Barakmor1** [comment]
@dankenigsberg @dasionov Splitting bug fixes from new capabilities is important, as it makes backporting easier, at least from my perspective. We should avoid including new capabilities in backports.

Please consider this before proceeding here.

File: pkg/virt-handler/node-labeller/arm64.go [hunk 2]
```diff
@@ -33,3 +37,10 @@ func (archLabellerARM64) arch() string {
 func (archLabellerARM64) shouldLabelNodes() bool {
 	return true
 }
+
+func (archLabellerARM64) prepareLabels(n *NodeLabeller, _ *v1.Node, _ []string, _ cpuFeatures, _ hostCPUModel, _ map[string]bool) map[string]string {
+	newLabels := make(map[string]string)
+	n.prepareMachineTypeLabels(newLabels)
```
Comment (iholder101) [inline pkg/virt-handler/node-labeller/arm64.go:43]: So the NodeLabeller calls the archLabeller, which calls the NodeLabeller again to perpare the labels? This looks off.

I'd much rather to add methods such as `AddNamedCpuLabels()` or `AddMachineTypeLabels()`.

Comment (dasionov) [reply pkg/virt-handler/node-labeller/arm64.go:43]: yes, this part can be refactored a little bit.. I mainly wanted to reduce duplication between the defaultArchLabeller and the Arm64 one

**Comment by iholder101** [comment]
@dasionov I agree with @Barakmor1's comments above.

What this PR is trying to achieve?
If this is an enhancement which seeks to advertise the supported named models for ARM, then:
1. I'd be happy to understand why it wasn't enabled in the past (e.g. if there was some libvirt limitation that is now gone?) and why we're now ready for it.
2. We need this to be properly tested upstream. Unfortunately, [ARM lanes are being dropped](https://groups.google.com/g/kubevirt-dev/c/WfW7hTWxJyQ), but we can still maybe leverage KubevirtCI for this job.
3. It will be hard to justify a backport.

If this is a bug fix that seeks to address machine-type node compatibility problem, let's tackle the minimum necessary, and advertise only whatever we have to in order to fix the bug.

File: pkg/virt-handler/node-labeller/arm64.go (hunk 2 above)
Comment (iholder101) [reply pkg/virt-handler/node-labeller/arm64.go:43]: > yes, this part can be refactored a little bit.. I mainly wanted to reduce duplication between the defaultArchLabeller and the Arm64 one

Sounds good.
I think you can create methods that decide what to expose and what not to expose, but keep this logic at the node-labeller level.

**Comment by dasionov** [comment]
> Hey @dasionov , could you please update the description to clarify exactly which labels we want to support for ARM, why these labels weren’t supported before, and why it’s now possible to support them? The current description doesn’t seem up to date.
> 
> Also, I suggest creating a parallel PR to update the [user guide](https://kubevirt.io/user-guide/cluster_admin/virtual_machines_on_Arm64/#virtual-machines-on-arm64) with these changes
> 
> By the way, do we have any documentation for the new machine type labels?

I Updated the description as you suggested :pray:

**Comment by iholder101** [comment]
> > Hey @dasionov , could you please update the description to clarify exactly which labels we want to support for ARM, why these labels weren’t supported before, and why it’s now possible to support them? The current description doesn’t seem up to date.
> > Also, I suggest creating a parallel PR to update the [user guide](https://kubevirt.io/user-guide/cluster_admin/virtual_machines_on_Arm64/#virtual-machines-on-arm64) with these changes
> > By the way, do we have any documentation for the new machine type labels?
> 
> I Updated the description as you suggested 🙏

Thanks a lot @dasionov! Looks much better!

> Previously, the node-labeller was disabled for ARM64 clusters due to incomplete support for labels like CPU features and models, which are supported on other architectures such as amd64 and s390x.

Can you please either provide a reference for this statement (or remove it)? Was this a libvirt limitation that had been addressed?

**Comment by dasionov** [comment]
> > Previously, the node-labeller was disabled for ARM64 clusters due to incomplete support for labels like CPU features and models, which are supported on other architectures such as amd64 and s390x.
> 
> Can you please either provide a reference for this statement (or remove it)? Was this a libvirt limitation that had been addressed?

I opened a bug here https://issues.redhat.com/browse/RHEL-88219

File: pkg/virt-handler/node-labeller/arch_labeller.go [hunk 3]
```diff
@@ -81,3 +87,64 @@ func (defaultArchLabeller) supportsHostModel() bool {
 func (defaultArchLabeller) arch() string {
 	return runtime.GOARCH
 }
+
+// prepareLabels converts cpu models, features, hyperv features to map[string]string format
+// e.g. "cpu-feature.node.kubevirt.io/Penryn": "true"
+func (defaultArchLabeller) prepareLabels(nodeLabeller *NodeLabeller, node *v1.Node, cpuModels []string, cpuFeatures cpuFeatures, hostCpuModel hostCPUModel, obsoleteCPUsx86 map[string]bool) map[string]string {
```
Comment (0xFelix) [inline pkg/virt-handler/node-labeller/arch_labeller.go:93]:
```suggestion
func (nodeLabeller *defaultArchLabeller) prepareLabels(node *v1.Node, cpuModels []string, cpuFeatures cpuFeatures, hostCpuModel hostCPUModel, obsoleteCPUsx86 map[string]bool) map[string]string {
```

Pardon my ignorance, something like this won't work?

File: pkg/virt-handler/node-labeller/node_labeller_test.go [hunk 4]
```diff
@@ -91,7 +93,7 @@ var _ = Describe("Node-labeller ", func() {
 					MinCPUModel:       "Penryn",
 				},
 			},
-		})
+		}, "testdata", runtime.GOARCH)
```
Comment (0xFelix) [inline pkg/virt-handler/node-labeller/node_labeller_test.go:96]: Can you make this independent of the host running the unit tests? I.e. do not use `runtime.GOARCH` but consts with the arches instead?

File: pkg/virt-handler/node-labeller/arch_labeller.go [hunk 5]
```diff
@@ -41,6 +46,7 @@ type archLabeller interface {
 	hasHostSupportedFeatures() bool
 	supportsHostModel() bool
 	arch() string
+	prepareLabels(n *NodeLabeller, node *v1.Node, cpuModels []string, cpuFeatures cpuFeatures, hostCpuModel hostCPUModel, obsoleteCPUsx86 map[string]bool) map[string]string
```
Comment (Barakmor1) [inline pkg/virt-handler/node-labeller/arch_labeller.go:49]: I guess that `	shouldLabelNodes() bool` and [this](https://github.com/kubevirt/kubevirt/blob/d0dd376266d025ff02e612a1cc733526904dc77d/cmd/virt-handler/virt-handler.go#L317) if statement are not needed anymore right?

File: pkg/virt-handler/node-labeller/arch_labeller.go (hunk 3 above)
Comment (dasionov) [reply pkg/virt-handler/node-labeller/arch_labeller.go:93]: The `NodeLabeller` has a reference to the `archLabeller` (which is basically a configuration of what is supported on each arch), and also it holds a lot of data related to the labels it needs to handle, they are not the same object

i agree about the fact that it looks weird, passing a reference to the original `NodeLabeller` into a method which is reside in the `archLabeller` now, but what alternative is there, the original method needs access to the data of the `NodeLabeller` (machine-types, cpu vendor, host-model etc..), and moving the `prepareLabels` seems like a good engineering because we want to support specific stuff when the arch is different then the default.

File: pkg/virt-handler/node-labeller/node_labeller_test.go (hunk 4 above)
Comment (dasionov) [reply pkg/virt-handler/node-labeller/node_labeller_test.go]: fixed, using `amd64` for the default tests.

File: pkg/virt-handler/node-labeller/arm64.go (hunk 2 above)
Comment (iholder101) [reply pkg/virt-handler/node-labeller/arm64.go]: > I'd much rather to add methods such as AddNamedCpuLabels() or AddMachineTypeLabels().

I see that the recent changes didn't include these.
Do you agree it's the path forward?

File: cmd/virt-handler/virt-handler.go [hunk 6]
```diff
@@ -314,11 +314,9 @@ func (app *virtHandlerApp) Run() {
 		panic(err)
 	}
 
-	if nodeLabellerController.ShouldLabelNodes() {
-		hostCpuModel = nodeLabellerController.GetHostCpuModel().Name
+	hostCpuModel = nodeLabellerController.GetHostCpuModel().Name
```
Comment (xpivarc) [inline cmd/virt-handler/virt-handler.go:317]: The af31c3e9071b665cc0cf6728af2df3ecc2d2e150 latest commit could be squashed to the first commit (or the first one can simply be removed)

**Comment by xpivarc** [comment]
/cherry-pick release-1.5

As alternative we can patch https://github.com/kubevirt/kubevirt/pull/13690 to be x86 specific for now but I would prefer this PR over the approach.

File: pkg/virt-handler/node-labeller/arch_labeller.go (hunk 3 above)
Comment (0xFelix) [reply pkg/virt-handler/node-labeller/arch_labeller.go:88]: I see

File: pkg/virt-handler/node-labeller/node_labeller_test.go [hunk 7]
```diff
@@ -298,6 +300,58 @@ var _ = Describe("Node-labeller ", func() {
...
+
+			for _, guest := range guestsCaps {
+				for _, machine := range guest.Arch.Machines {
+					expectedLabelKey := v1.SupportedMachineTypeLabel + machine.Name
+					Expect(node.Labels).To(HaveKey(expectedLabelKey), "expected machine type label %s to be present", expectedLabelKey)
+				}
+			}
+		})
+
+		It("should not add invalid labels", func() {
```
[the remark this answers is not in the document]
Comment (dasionov) [reply pkg/virt-handler/node-labeller/node_labeller_test.go:324]: so should i drop it ?

**Comment by zhlhahaha** [comment]
Sorry for the delayed reply — it seems there's quite a bit of context around this PR. I’ve just gone through it.

I’m not very familiar with the code in nodelabeller, so I’ll just share some thoughts regarding live migration of VMs on the Arm64 platform.

As @andreabolognani mentioned, host-model is not supported on Arm64. Unlike x86 CPUs (manufactured by Intel and AMD) or s390x (by IBM), Arm64 CPUs are produced by a wide range of vendors such as AWS, NVIDIA, Ampere, Alibaba, and others. Each vendor may introduce custom features or designs in their CPUs, which makes cross-vendor live migration impractical.

In practice, we only migrate VMs between hosts using the same Arm64 CPU model from the same vendor. Based on this, labeling nodes by vendor and CPU version — rather than by specific CPU features — would be more appropriate for Arm64.

However, collecting vendor and CPU version information reliably is challenging. I’ve tried using lscpu, but on some machines, the output lacks this data. Given these constraints, I think it might be reasonable (at least for now) to either use a placeholder or find a way to bypass the node labeller on Arm64 altogether.

File: cmd/virt-handler/virt-handler.go (hunk 6 above)
Comment (dasionov) [reply cmd/virt-handler/virt-handler.go:317]: right, dropped the first commit

File: pkg/virt-handler/node-labeller/arm64.go [hunk 8]
```diff
@@ -29,3 +33,10 @@ type archLabellerARM64 struct {
 func (archLabellerARM64) arch() string {
 	return arm64
 }
+
+func (archLabellerARM64) prepareLabels(nodeLabeller *NodeLabeller, _ *v1.Node) map[string]string {
+	newLabels := make(map[string]string)
+	addMachineTypeLabels(nodeLabeller, newLabels)
+
+	return newLabels
+}
```
Comment (iholder101) [inline pkg/virt-handler/node-labeller/arm64.go:42]: Further to https://github.com/kubevirt/kubevirt/pull/14520#discussion_r2056264520, I don't think we should extract `prepareLabels` to the arch labeller. Instead, we can ask the arch labeller whether we should advertise named models and so on.

something like:
```go
if n.arch.supportsNamedModels {
	// advertise labels...
}
if n.arch.supportsMachineType {
	// ...
}
```

File: pkg/virt-handler/node-labeller/arm64.go (hunk 2 above)
Comment (iholder101) [reply pkg/virt-handler/node-labeller/arm64.go]: That's not what I meant exactly :)
See https://github.com/kubevirt/kubevirt/pull/14520/files#r2058690100

File: pkg/virt-handler/node-labeller/node_labeller_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (xpivarc) [reply pkg/virt-handler/node-labeller/node_labeller_test.go]: Yes please, unless you don't agree with me and can reason why we need it

**Comment by xpivarc** [comment]
@zhlhahaha Thanks! 
This is exactly what we aim for now, We want to enable live migration for homogeneous clusters first. We asked Libvirt project to help with the heterogeneous case.

File: pkg/virt-handler/node-labeller/node_labeller.go [hunk 9]
```diff
@@ -235,27 +230,36 @@ func (n *NodeLabeller) loadHypervFeatures() {
...
 	newLabels := make(map[string]string)
-	for key := range cpuFeatures {
-		newLabels[kubevirtv1.CPUFeatureLabel+key] = "true"
-	}
 
-	for _, value := range cpuModels {
-		newLabels[kubevirtv1.CPUModelLabel+value] = "true"
-		newLabels[kubevirtv1.SupportedHostModelMigrationCPU+value] = "true"
+	if n.arch.hasHostSupportedFeatures() {
+		for key := range cpuFeatures {
```
Comment (xpivarc) [inline pkg/virt-handler/node-labeller/node_labeller.go:242]: nit: You can call `n.getSupportedCpuFeatures()` directly

File: pkg/virt-handler/node-labeller/node_labeller.go [hunk 10]
```diff
@@ -235,27 +230,36 @@ func (n *NodeLabeller) loadHypervFeatures() {
...
+			newLabels[kubevirtv1.CPUFeatureLabel+key] = "true"
+		}
 	}
 
-	// Add labels for supported machine types
-	machines := n.getSupportedMachines()
+	if n.arch.supportsNamedModels() {
+		for _, value := range cpuModels {
```
Comment (xpivarc) [inline pkg/virt-handler/node-labeller/node_labeller.go:248]: nit: `n.getSupportedCpuModels(obsoleteCPUsx86)`

File: pkg/virt-handler/node-labeller/node_labeller_test.go [hunk 11]
```diff
@@ -298,6 +312,47 @@ var _ = Describe("Node-labeller ", func() {
...
+			}, "testdata/arm64", arm64)
+			mockQueue := testutils.NewMockWorkQueue(nlController.queue)
+			nlController.queue = mockQueue
+
+			mockQueue.ExpectAdds(1)
+			nlController.queue.Add(nodeName)
+			mockQueue.Wait()
+		})
+
+		It("should add machine type label", func() {
```
[the remark this answers is not in the document]
Comment (dasionov) [reply pkg/virt-handler/node-labeller/node_labeller_test.go]: done

File: pkg/virt-handler/node-labeller/node_labeller.go (hunk 9 above)
Comment (dasionov) [reply pkg/virt-handler/node-labeller/node_labeller.go]: done

File: pkg/virt-handler/node-labeller/node_labeller.go (hunk 10 above)
Comment (dasionov) [reply pkg/virt-handler/node-labeller/node_labeller.go]: done

File: pkg/virt-handler/node-labeller/node_labeller_test.go [hunk 12]
```diff
@@ -112,6 +112,20 @@ var _ = Describe("Node-labeller ", func() {
 		Expect(nlController.queue.Len()).To(BeZero(), "labeller should process all nodes from queue")
 	})
 
+	It("should add machine type labels", func() {
```
[the remark this answers is not in the document]
Comment (dasionov) [reply pkg/virt-handler/node-labeller/node_labeller_test.go]: done

File: pkg/virt-handler/node-labeller/arm64.go (hunk 8 above)
Comment (dasionov) [reply pkg/virt-handler/node-labeller/arm64.go]: added `supportsNamedModels()`, `supportsMachineType()` should be supported for all if we are not patching up #13690 for amd only.

**Review by xpivarc — COMMENTED** [review]
/approve

File: pkg/virt-handler/node-labeller/node_labeller.go [hunk 13]
```diff
@@ -235,27 +230,33 @@ func (n *NodeLabeller) loadHypervFeatures() {
...
 		labelKey := kubevirtv1.SupportedMachineTypeLabel + machine.Name
 		newLabels[labelKey] = "true"
 	}
 
-	if _, hostModelObsolete := obsoleteCPUsx86[hostCpuModel.Name]; !hostModelObsolete {
-		newLabels[kubevirtv1.SupportedHostModelMigrationCPU+hostCpuModel.Name] = "true"
+	if n.arch.supportsHostModel() {
+		if _, hostModelObsolete := obsoleteCPUsx86[hostCpuModel.Name]; !hostModelObsolete {
+			newLabels[kubevirtv1.SupportedHostModelMigrationCPU+hostCpuModel.Name] = "true"
+		}
```
Comment (Barakmor1) [inline pkg/virt-handler/node-labeller/node_labeller.go:259]: nit: this can be move to the if statement below

**Review by fossedihelm — COMMENTED** [review]
Thank you!
/lgtm

File: pkg/virt-handler/node-labeller/node_labeller.go (hunk 13 above)
Comment (dasionov) [reply pkg/virt-handler/node-labeller/node_labeller.go]: done!

File: pkg/virt-handler/node-labeller/node_labeller_test.go [hunk 14]
```diff
@@ -289,15 +289,31 @@ var _ = Describe("Node-labeller ", func() {
...
+		}
 	},
-		Entry(amd64, amd64, true),
-		Entry(arm64, arm64, false),
-		Entry(s390x, s390x, true),
-		Entry("unknown", "unknown", false),
+		Entry("for amd64", []libvirtxml.CapsGuestMachine{{Name: "q35"}, {Name: "q35-rhel9.6.0"}}),
+		Entry("for arm64", []libvirtxml.CapsGuestMachine{{Name: "virt"}, {Name: "virt-rhel9.6.0"}}),
 	)
+
```
[the remark this answers is not in the document]
Comment (dasionov) [reply pkg/virt-handler/node-labeller/node_labeller_test.go]: done

File: pkg/virt-handler/node-labeller/node_labeller.go [hunk 15]
```diff
@@ -267,19 +262,25 @@ func (n *NodeLabeller) prepareLabels(node *v1.Node, cpuModels []string, cpuFeatu
...
+			newLabels[kubevirtv1.SupportedHostModelMigrationCPU+hostCpuModel.Name] = "true"
 		}
-	}
 
-	newLabels[kubevirtv1.CPUModelVendorLabel+n.cpuModelVendor] = "true"
-	newLabels[kubevirtv1.HostModelCPULabel+hostCpuModel.Name] = "true"
+		for feature := range hostCpuModel.requiredFeatures {
+			newLabels[kubevirtv1.HostModelRequiredFeaturesLabel+feature] = "true"
+		}
+		if _, obsolete := obsoleteCPUsx86[hostCpuModel.Name]; obsolete {
```
[the remark this answers is not in the document]
Comment (dasionov) [reply pkg/virt-handler/node-labeller/node_labeller.go:273]: done

**Comment by fossedihelm** [comment]
Thank you!
